lib/src/contracts_generator_config.dart

```diff
@@ -21,18 +21,74 @@ class ContractsGeneratorConfig {
 
     final config = mergeMaps<dynamic, dynamic>(defaultConfig, override);
 
+    late GeneratorScript script;
+
     if (config['input'] == null) {
       throw ArgumentError.notNull(
         '`input` is a required field and cannot be null',
       );
+    } else {
```

Comment:
A slight refactoring would be beneficial here, e.g. extracting a function for each case & changing the validation to something more direct.

Reply:
> changing the validation to something more direct.

Could you elaborate?

Reply:
So, now you have validation & parsing intertwined. This is quite hard to follow and test thoroughly. It is much easier to write (and maintain) validation as a part of parsing, i.e. you have three methods that each considers only a single _way_ of configuration. Doing _magic_ on a map, with `length` check, is rather _magic_ and quite hard to follow and extend.

To visualise what I am talking about, see this pseudocode:
```dart
GeneratorScript configureProject(dynamic config) {
  if (config['input']['project'] != null) {
    return GeneratorScript.project(config['input']['project'].value);
  } else {
    return null;
  }
}

// others look alike

GeneratorScript configure(dynamic config) {
  final proj = configureProject(config);
  final file = configureFile(config);
  final path = configurePath(config);
  
  // This check should be handled by the argument parsing library, not you
  if (proj != null ^ file != null ^ path != null) {
    throw ArgumentException('invalid config');
  }

  return proj ?? file ?? path;
}

// And the `fromYaml`
factory ContractsGeneratorConfig.fromYaml(String yaml) {
    final defaultConfig = loadYaml(defaultYamlConfig) as Map;
    final override = loadYaml(yaml) as Map;

    final config = mergeMaps<dynamic, dynamic>(defaultConfig, override);
    final dynamic name = config['name'];
    final dynamic include = config['include'];
    final dynamic output = config['output'];
    final dynamic extra = config['extra'];

    if (name is! String) {
      throw ArgumentError('`name` field has to be a string');
    }
    if (include is! String) {
      throw ArgumentError('`include` field has to be a string');
    }
    if (output is! String) {
      throw ArgumentError('`output` field has to be a string');
    }
    if (extra is! String) {
      throw ArgumentError('`extra` field has to be a string');
    }

    return ContractsGeneratorConfig(
      input: configure(config),
      name: name,
      include: RegExp(include),
      output: Directory(output),
      extra: extra,
    );
}
```